rehttp_server_post17_test.go

```diff
@@ -0,0 +1,136 @@
+			})),
+		ConstDelay(time.Duration(0)),
+	)
+	tr.PerAttemptTimeout = time.Millisecond * 10 // short timeout
+
+	client := http.Client{
+		Transport: tr,
+	}
+
+	req, err := http.NewRequest(http.MethodGet, ts.URL, nil)
+	if err != nil {
+		t.Errorf("error creating request: %s", err)
+	}
+
+	_, err = client.Do(req)
+	time.Sleep(time.Millisecond * 100) // let the server finish sleeping
+	// should have attempted 4 times but errored due to timeout
+	attemptCountMu.Lock()
+	if attemptCount != 4 {
+		t.Errorf("attempt count does not match expected: got %d, want %d", attemptCount, 4)
+	}
+	if err == nil {
+		t.Error("expected timeout error doing request but got nil")
+	}
+
+	attemptCount = 0
+	attemptCountMu.Unlock()
+
+	// now increase the timeout restriction
+	tr.PerAttemptTimeout = time.Second
+	res, err := client.Do(req)
+	attemptCountMu.Lock()
+	// should have attempted 4 times without going over the timeout
+	if attemptCount != 4 {
+		t.Errorf("attempt count does not match expected: got %d, want %d", attemptCount, 4)
+	}
+	if err != nil {
+		t.Errorf("got unexpected error doing request: %s", err)
+	}
+	if res == nil || res.StatusCode != http.StatusTooManyRequests {
+		t.Errorf("status code does not match expected: got %d, want %d", res.StatusCode, http.StatusTooManyRequests)
+	}
+
+	attemptCount = 0
+	attemptCountMu.Unlock()
+
+	// now remove the timeout restriction
+	tr.PerAttemptTimeout = time.Duration(0)
+	res, err = client.Do(req)
+	attemptCountMu.Lock()
+	// should have attempted 4 times without going over the timeout
+	if attemptCount != 4 {
+		t.Errorf("attempt count does not match expected: got %d, want %d", attemptCount, 4)
+	}
+	if err != nil {
+		t.Errorf("got unexpected error doing request: %s", err)
+	}
+	if res == nil || res.StatusCode != http.StatusTooManyRequests {
+		t.Errorf("status code does not match expected: got %d, want %d", res.StatusCode, http.StatusTooManyRequests)
+	}
+	attemptCountMu.Unlock()
+}
+
+// TestCancelReader is meant to test that the cancel reader is correctly
+// preventing the race-case of being unable to read the body due to a
+// preemptively-canceled context.
+func TestCancelReader(t *testing.T) {
+	rt := NewTransport(http.DefaultTransport, RetryMaxRetries(1), ConstDelay(time.Duration(0)))
+	rt.PerAttemptTimeout = time.Millisecond * 100
+	client := http.Client{
+		Transport: rt,
+	}
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(time.Millisecond * 10)
+		w.WriteHeader(http.StatusOK)
+		// need a decent number of bytes to make the race case more likely to fail
+		// https://github.com/go-kit/kit/issues/773
+		w.Write(make([]byte, 102400))
+	}))
+	defer ts.Close()
+
+	ctx := context.Background()
+
+	req, _ := http.NewRequest(http.MethodGet, ts.URL, nil)
+	res, err := client.Do(req.WithContext(ctx))
+	if err != nil {
+		t.Errorf("unexpected error creating request: %s", err)
+	}
+	if res == nil {
+		t.Error("unexpected nil response")
+	}
```

Comment:
I don't think `client.Do` will ever return `nil` if `err != nil`? I think those `res==nil` or `res != nil` checks can savely be removed.